File: cpp/problems/leetcode/0011_container_with_most_water/solution.cpp
// LeetCode 11
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0, r = static_cast<int>(height.size()) - 1, best = 0;
        while (l < r) {
            int h = min(height[l], height[r]);
            best = max(best, h * (r - l));
            if (height[l] < height[r])
                ++l;
            else
                --r;
        }
        return best;
    }
};

int main() {
    vector<int> h{1, 8, 6, 2, 5, 4, 8, 3, 7};
    assert(Solution().maxArea(h) == 49);
    cout << "0011 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0011_container_with_most_water/solution.py
"""LeetCode 11: Container With Most Water."""

from __future__ import annotations


class Solution:
    def maxArea(self, height: list[int]) -> int:
        l, r = 0, len(height) - 1
        best = 0
        while l < r:
            h = min(height[l], height[r])
            best = max(best, h * (r - l))
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return best


if __name__ == "__main__":
    assert Solution().maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]) == 49
    print("0011 OK")
